strategy_simulator/device.py

```diff
@@ -0,0 +1,311 @@
+from .messages import ChunkDescriptor, AnyMessage, DataMessage, AnnounceMessage, RequestMessage, Proto, RawData, FWType, \
+    Version
+from .iqueue import WriteQueue, ReadQueue
+from .clock import ClockView
+from .rs_store import RecentlySeenStore, RequestStore
+
+DeviceId = int
+DeviceType = int
+
+
+class OngoingUpgrade:
+    def __init__(self, fw_type: FWType, version: Version, proto: Proto):
+        self.fw_type: FWType = fw_type
+        self.version: Version = version
+        self.proto: Proto = proto
+        self.last_progress: int = -1
+        self.candidate_firmware: Firmware = Firmware(fw_type, version, [None] * proto.chunks)
+
+
+class Device:
+    CHUNK_SIZE: int = 1
+
+    # limitation: all neighbors share the same queue for the target device thus we're not able to model each link's
+    # reliability alone thus reliability should, from this PoV, be a property of the device
+    def __init__(self, dev_id: DeviceId, dev_type: DeviceType, input_queue: ReadQueue,
+                 neighbors: Dict[DeviceId, WriteQueue], running_firmware: Firmware, clock: ClockView,
+                 diff_announces_seen_store=None, in_flight_requests_store=None, datas_seen_store=None):
+        self.dev_id: DeviceId = dev_id
+        self.dev_type: DeviceType = dev_type
+
+        self._clock = clock
+        self._stats: Dict[Any, Any] = {}  # temporary solution, to be moved to simulator
+        self._input_queue: ReadQueue = input_queue
+        self.neighbors: Dict[DeviceId, WriteQueue] = neighbors
+
+        self.running_firmware: Firmware = running_firmware
+
+        self._ongoing_upgrade: Optional[OngoingUpgrade] = None
+        self._current_message: Optional[AnyMessage] = None  # todo: remove this hacky way
+
+        self.periodic_announce: int = 100
+        self._last_periodic_announce: int = -self.periodic_announce
+        # todo: if self.in_flight_timeout >= self.progress_timeout then spontaneous requests do not work
+        #  either check this condition or make self.upgrade_process_timeout_handler
+        #  use unrestricted impl of _request_message_for that would send all requests without limits/restrictions
+        self.progress_timeout: int = 100  # todo: pass from argument
+        # self._last_progress resides inside self._ongoing_upgrade
+
+        self._diff_announces_seen_store = diff_announces_seen_store or RecentlySeenStore(self._clock, timeout=self.periodic_announce//2, max_capacity=None)
+        self._in_flight_requests_store = in_flight_requests_store or RequestStore(self._clock, timeout=self.progress_timeout//2, max_capacity=None)
+        self._datas_seen_store = datas_seen_store or RecentlySeenStore(self._clock, timeout=self.progress_timeout//2, max_capacity=None)
+
+    @property
+    def upgrading(self) -> bool:
+        """The device is being upgraded/is in a process of upgrade of its firmware"""
+        return self._ongoing_upgrade is not None
+
+    def on_before_message(self, m: AnyMessage) -> bool:
```

Comment:
Není toto divné jméno funkce? Podle názvu je to logika, která se musí stát před každým zpracováním zprávy, ale ve skutečnosti mi to připadá jako samotné zpracování zprávy. Zároveň nějaké zpracování zprávy se děje v místě volání `on_before_message` a je to takové "chlupaté".